### PR TITLE
Match functions

### PR DESCRIPTION
**Pull request by abargar**

### CODE REVIEW

File: modules/matcher.py [hunk 1]
```diff
@@ -24,11 +24,10 @@
...
 
-def basic_preprocess(df: pd.DataFrame, feature: str) -> pd.DataFrame:
-    df = df[[DOMAIN, feature]]
-    df = df[~df[feature].isna() & ~df[feature].isnull()]
-
-    return df
+def basic_preprocess(df: pd.DataFrame) -> pd.DataFrame:
+    df = df[[DOMAIN, INDICATOR]]
+    df = df[~df[INDICATOR].isna() & ~df[INDICATOR].isnull()]
+    return df.drop_duplicates()
```
Comment (abargar) [inline modules/matcher.py:30]: remove duplicate features

File: tests/test__matcher.py [hunk 2]
```diff
@@ -113,20 +111,17 @@ def test__direct_match(feature_df, compare_df, expected_results):
                     {
                         "domain_name_x": "a",
                         "domain_name_y": "b",
-                        "match_type": "feature",
-                        "match_value": 0.2,
+                        "match_value": False,
```
Comment (abargar) [inline tests/test__matcher.py:114]: these will be updated back to the numeric values

File: modules/matcher.py [hunk 3]
```diff
@@ -37,22 +38,30 @@ def column_contains_list_string(column: pd.Series) -> bool:
...
         df_copy[INDICATOR] = df_copy[INDICATOR].map(ast.literal_eval)
-        return df_copy.groupby(DOMAIN)[INDICATOR].agg(lambda x: set(chain.from_iterable(x)))
+        result = df_copy.groupby(DOMAIN)[INDICATOR].agg(
+            lambda x: set(chain.from_iterable(x))
+        )
     else:
-        return df.groupby(DOMAIN)[INDICATOR].apply(set)
-
+        result = df.groupby(DOMAIN)[INDICATOR].apply(set)
+    return result[result.str.len() > 0]
```
Comment (abargar) [inline modules/matcher.py:64]: ensure we remove all features where the group of indicators is empty ([] or {})

File: modules/matcher.py [hunk 4]
```diff
@@ -113,85 +122,97 @@ def cert_preprocess(df: pd.DataFrame, cert_feature: str) -> pd.DataFrame:
...
+    matches = matches.rename(columns={INDICATOR: MATCH_VALUE})
     return matches.reset_index(drop=True)
 
-# TODO: Add a partial string match function
+def match_with_threshold(
+        feature_df,
+        comparison_df, 
+        match_function,
+        threshold
+):
```
Comment (abargar) [inline modules/matcher.py:137]: this function handles the comparison of the two dataframes, matching, and thresholding. This way the other functions only have to handle the definition of the `match_function`

File: modules/matcher.py [hunk 5]
```diff
@@ -113,85 +122,97 @@ def cert_preprocess(df: pd.DataFrame, cert_feature: str) -> pd.DataFrame:
...
+    result = pd.DataFrame(
+        match_data, columns=["domain_name_x", "domain_name_y", MATCH_VALUE]
+    ).drop_duplicates(["domain_name_x", "domain_name_y"])
+    return result
+
+def partial_text_match(
+    feature_df: pd.DataFrame,
+    comparison_df: pd.DataFrame,
+    threshold: float = 0.9,
+) -> pd.DataFrame:
```
Comment (abargar) [inline modules/matcher.py:164]: function for partial text matching

File: modules/matcher.py [hunk 6]
```diff
@@ -299,7 +317,7 @@ def parse_certificate_matches(
 "2-urlscan_urlssonpage" : iou_match,
 "2-urlscanhrefs" : iou_match,
 "2-techstack" : iou_match,
-"3-footer-text": direct_match,
+"3-footer-text": partial_text_match,
```
Comment (abargar) [inline modules/matcher.py:320]: footer now is matched with partial text matching

File: tests/test__matcher.py [hunk 7]
```diff
@@ -350,24 +297,17 @@ def test__main(
...
     output_file,
 ):
     mock_define_output_filename.return_value = "r_file"
     mock_read_csv.return_value = "fake_data"
-    main(input_file, compare_file, result_dir, output_file, comparison_type)
-    if result_dir:
-        mock_mkdir.assert_called_once_with(exist_ok=True)
-    else:
-        mock_mkdir.assert_not_called()
+    main(input_file, compare_file, output_file, comparison_type)
```
Comment (abargar) [inline tests/test__matcher.py:304]: the intermediate result directory doesn't make sense in a web application. I'm going to move this script out of here so you can call it like a normal script. If you want the result directory with it, I can add it back in.

**Review by pbenzoni — APPROVED** [review]

**Pull request merged**